feat: Add feature for copying backups

danieljbruce opened the pull request:
This PR includes changes in the veneer layer that implement the new copy backup feature which allows users to copy a backup of a table to a cluster of their choice.

A config interface is introduced which is used to specify details about the copied backup. On a backup object, a copy function is added which copies the backup to a destination specified by the config passed into the function parameter. Code that parses expiry time is reused for the copy backup function.

Unit tests are added to make sure source and destination information get passed along to the Gapic layer. System tests are added to check that copying a backup works when copying to the same cluster, different cluster and different project. System tests are also added to be sure that expiry times can be passed in using different formats and tests are added so that we can confirm restoring the copied backup works properly.

danieljbruce committed:
src test and protos copied over

danieljbruce committed:
lint fix

danieljbruce committed:
implemented copy backup

danieljbruce committed:
part of test done

danieljbruce committed:
workaround

danieljbruce committed:
corrections to get code working

danieljbruce committed:
Add expire time

danieljbruce committed:
This allows the copy backup endpoint to work

danieljbruce committed:
Working system test

danieljbruce committed:
delete try block

danieljbruce committed:
tests pass with new refactor

danieljbruce committed:
Create backup and copy on another cluster

danieljbruce committed:
modifications to second test

danieljbruce committed:
copy backup unit test

danieljbruce committed:
Revert "lint fix"

This reverts commit 576550d404d249430d1495602831926f74ef3bbb.

danieljbruce committed:
Revert "src test and protos copied over"

This reverts commit 46ce5b8009a001b1197fbf9bbdded805a7ba3def.

danieljbruce committed:
First correction

kolea2 commented on system-test/bigtable.ts:
```diff
@@ -1417,6 +1417,66 @@ describe('Bigtable', () => {
       Object.keys(policy).forEach(key => assert(key in updatedPolicy));
     });
+    describe('copying backups', () => {
+      async function testCopyBackup(backup: Backup, newBackup: Backup) {
```
I would like to see some more assertions here. I am thinking something along the lines of:
1. copy the backup, wait for operation to complete
2. list backups
3. assert the copied backup is in the list, and assert it has correct name, expire time, etc

danieljbruce replied:
list backups contains name and id so I made an assert that compares those two and ensures exactly one new backup gets created that matches.

kolea2 commented on test/backup.ts:
```diff
@@ -223,6 +224,38 @@ describe('Bigtable/Backup', () => {
+    it.only('should correctly copy backup from the Cluster', done => {
+      const backupId = generateId('backup');
+      const newBackupId = generateId('backup');
+      const oldBackup = new Backup(CLUSTER, backupId);
```
nit - for clarity, I would rename these `backup` and `copiedBackup`

danieljbruce replied:
Done

danieljbruce committed:
Check the list of backups

danieljbruce committed:
fetched backup

danieljbruce committed:
get name and id

danieljbruce committed:
Remove only

danieljbruce committed:
refactor request mock

danieljbruce committed:
Add unit test for copying backup to another projec

danieljbruce committed:
Merge branch 'main' into copy-backup-2-relevant-changes-only

gcf-owl-bot[bot] committed:
feat: add experimental reverse scan for public preview

PiperOrigin-RevId: 543539118

Source-Link: https://github.com/googleapis/googleapis/commit/ae187063e3d8a43d85edb9b3084413d568ce7945

Source-Link: https://github.com/googleapis/googleapis-gen/commit/5d05516f84e53aaba63a4b8767ff955ac5bb4a87
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiNWQwNTUxNmY4NGU1M2FhYmE2M2E0Yjg3NjdmZjk1NWFjNWJiNGE4NyJ9

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
feat: Increase the maximum retention period for a Cloud Bigtable backup from 30 days to 90 days

PiperOrigin-RevId: 544356969

Source-Link: https://github.com/googleapis/googleapis/commit/c35889a0b917e22e26c53acafa5c27102a51d623

Source-Link: https://github.com/googleapis/googleapis-gen/commit/c00326ec78565b5d16f92c845ff0bb18f11ca05d
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiYzAwMzI2ZWM3ODU2NWI1ZDE2ZjkyYzg0NWZmMGJiMThmMTFjYTA1ZCJ9

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
docs: fix formatting for reversed order field example

PiperOrigin-RevId: 547553954

Source-Link: https://github.com/googleapis/googleapis/commit/c4e6427fcefd1cd9a15a3008ae7ee8adca972276

Source-Link: https://github.com/googleapis/googleapis-gen/commit/f552269609d4183546543bfe3a022f544d4f5bdb
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiZjU1MjI2OTYwOWQ0MTgzNTQ2NTQzYmZlM2EwMjJmNTQ0ZDRmNWJkYiJ9

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
feat: add last_scanned_row_key feature

PiperOrigin-RevId: 551191182

Source-Link: https://github.com/googleapis/googleapis/commit/51e04baa9eec3bee8b3e237bfd847eb06aa66d72

Source-Link: https://github.com/googleapis/googleapis-gen/commit/4b90e8ead4477eff96c31b9b0fdef36ed975b15f
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiNGI5MGU4ZWFkNDQ3N2VmZjk2YzMxYjliMGZkZWYzNmVkOTc1YjE1ZiJ9

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
fix: fix typings for IAM methods
docs: fixed links in the generated Markdown documentation

PiperOrigin-RevId: 551610576

Source-Link: https://github.com/googleapis/googleapis/commit/73b1313cbd1fd0cc1e22684bc89ee1b1a416cfe0

Source-Link: https://github.com/googleapis/googleapis-gen/commit/8bec066492a6da2855b1b8ce562664c0a6b30b01
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiOGJlYzA2NjQ5MmE2ZGEyODU1YjFiOGNlNTYyNjY0YzBhNmIzMGIwMSJ9

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
test: disable retry-request for streaming tests

PiperOrigin-RevId: 554648220

Source-Link: https://github.com/googleapis/googleapis/commit/53cd9ad1b48e40cdd44e0c13e96ac0281b32828f

Source-Link: https://github.com/googleapis/googleapis-gen/commit/7e8867efbed7dbfe5ef6ec3c2c92a4bce4280f7a
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiN2U4ODY3ZWZiZWQ3ZGJmZTVlZjZlYzNjMmM5MmE0YmNlNDI4MGY3YSJ9

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

sofisl committed:
Merge branch 'main' into owl-bot-copy

gcf-owl-bot[bot] committed:
feat: publish CopyBackup protos to external customers

PiperOrigin-RevId: 557192020

Source-Link: https://github.com/googleapis/googleapis/commit/b4c238feaa1097c53798ed77035bbfeb7fc72e96

Source-Link: https://github.com/googleapis/googleapis-gen/commit/feccb30e3177da8b7b7e68149ca4bb914f8faf2a
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiZmVjY2IzMGUzMTc3ZGE4YjdiN2U2ODE0OWNhNGJiOTE0ZjhmYWYyYSJ9

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
fix: simplify logic for HTTP/1.1 REST fallback option

For the `fallback` parameter, all values considered as `true`
in Boolean context will enable HTTP/1.1 REST fallback,
since the other fallback transport, proto over HTTP, is
removed from `google-gax` v4.

PiperOrigin-RevId: 559812260

Source-Link: https://github.com/googleapis/googleapis/commit/6a6fd29a79fe2846001d90d93e79a19fcc303b85

Source-Link: https://github.com/googleapis/googleapis-gen/commit/56c16657e7a59122b1da94771a9ef40989c282c0
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiNTZjMTY2NTdlN2E1OTEyMmIxZGE5NDc3MWE5ZWY0MDk4OWMyODJjMCJ9

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
fix: add feature flag for improved mutate rows throttling

PiperOrigin-RevId: 565090488

Source-Link: https://github.com/googleapis/googleapis/commit/e8a136feaca2547dd5566ef79841d28f76a80eb5

Source-Link: https://github.com/googleapis/googleapis-gen/commit/9a8dcca0fb2117628a1a6a6c3625a6aa32fc2f75
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiOWE4ZGNjYTBmYjIxMTc2MjhhMWE2YTZjMzYyNWE2YWEzMmZjMmY3NSJ9

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

sofisl committed:
Merge branch 'main' into owl-bot-copy

sofisl committed:
run lint

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

meredithslota committed:
Merge branch 'main' into copy-backup-2-relevant-changes-only

gcf-owl-bot[bot] committed:
build: update typescript generator version to publish in dual format (ESM)

PiperOrigin-RevId: 568643156

Source-Link: https://github.com/googleapis/googleapis/commit/f95afc063e20a0a61e13b186806ac84b49e329cf

Source-Link: https://github.com/googleapis/googleapis-gen/commit/bbd2c49d2e423a8ce5cc85627402d512aeefc58b
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiYmJkMmM0OWQyZTQyM2E4Y2U1Y2M4NTYyNzQwMmQ1MTJhZWVmYzU4YiJ9

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
feat: Add support for Cloud Bigtable Request Priorities in App Profiles

PiperOrigin-RevId: 571158646

Source-Link: https://github.com/googleapis/googleapis/commit/bc3c83b41b1589cca21f713a500f179ef86a7e18

Source-Link: https://github.com/googleapis/googleapis-gen/commit/93366e84e4e6861e2e580eb000721d99bf54a0a4
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiOTMzNjZlODRlNGU2ODYxZTJlNTgwZWIwMDA3MjFkOTliZjU0YTBhNCJ9

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

danieljbruce committed:
Merge branch 'main' into owl-bot-copy

gcf-owl-bot[bot] committed:
build: update Node.js generator to compile protos

PiperOrigin-RevId: 582493526

Source-Link: https://github.com/googleapis/googleapis/commit/7c4e4b52369c9f6ac3e78f945d36fc833f2280de

Source-Link: https://github.com/googleapis/googleapis-gen/commit/368cfb651016d6a93ca6e488cbc34e2d1d9d212c
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiMzY4Y2ZiNjUxMDE2ZDZhOTNjYTZlNDg4Y2JjMzRlMmQxZDlkMjEyYyJ9

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

danieljbruce committed:
Merge branch 'main' into copy-backup-2-relevant-changes-only

danieljbruce committed:
Merge branch 'owl-bot-copy' of https://github.com/googleapis/nodejs-bigtable into copy-backup-2-relevant-changes-only

danieljbruce committed:
Use the destination backup expiry time

Expiry time of the destination should be used for the call. Not for the source expiry time.

danieljbruce committed:
Use the config instead for copy backup

Use the config and not the backup object for copying backups. The expiry time is needed so a backup object can’t be used because the metadata cannot be introspected.

danieljbruce committed:
Rename CopyBackupConfig

Rename it to DestinationBackupConfig.

danieljbruce committed:
Merge branch 'main' into copy-backup-2-relevant-changes-only

danieljbruce committed:
Fix signature of copyBackup

Fix the signature of copyBackup to include the fact that the config now has the gax options.

danieljbruce committed:
Merge branch 'copy-backup-2-relevant-changes-only' of https://github.com/danieljbruce/nodejs-bigtable into copy-backup-2-relevant-changes-only

danieljbruce committed:
Change the copy signature to require callback

If the callback is not required then all calls match first signature overload and async calls return void which isn’t correct.

danieljbruce committed:
Some debugging on the system tests

The system tests were giving an error about the source expiry time and now this is fixed.

danieljbruce committed:
DestinationBackupConfig

Don’t use the name DestinationBackupConfig anymore.

danieljbruce committed:
lint fix

danieljbruce committed:
Move the config and create a generic callback

Create a generic callback data structure in order to avoid callback and promise data structures not matching up. Also move the configs to a different file.

danieljbruce committed:
Eliminate the generic callback data structure

Eliminate the generic data structure as it can’t be used to capture the pattern we need exactly. Also modify the way that errors are passed into the callback function to be more consistent.

danieljbruce committed:
Get the first copy backup test working

Copy the backup into the same cluster test is working. Still need to work on copy backup for test that looks at a different cluster to get that test working.

danieljbruce committed:
Fix different cluster, different instance test

Fix the issue preventing the second test from working by allowing the test function to specify the instance where the new backups should be created.

danieljbruce committed:
Add another test for copy backup

The test should address a backup copied to a different cluster, but the same instance.

danieljbruce committed:
Add a test case for copying to another project

Test case for copying to another project uses environment variable to point to new project for CI pipeline.

danieljbruce committed:
Restructure tests for multiple expire time inputs

Take one of the tests and structure it into a describe block instead. Each test in this block will now run the old test with a different expiry time format.

danieljbruce committed:
Various changes to test interaction with copy back

Rename backupId to id. id is a simpler name and it is the name that is used everywhere. Rename a function to testWithExpiryTimes. Modify source code for copy function to return a backup with the new id. Add test stub for restore backup.

danieljbruce committed:
Finish the restore tests and modify copy backup

The restore test should test to see if a table can be restored. It should also return a backup object that matches the new backup created.

danieljbruce committed:
Add comment describing second argument

Second argument is a backup that corresponds to a new backup. This needs to be explained in a comment.

danieljbruce committed:
Remove TODO

The TODO is done

danieljbruce committed:
Modify copy backup unit test

Modify the copy backup unit test to compare against copying the a destination with customized cluster, backup, instance and project.

danieljbruce committed:
Modify the name of the test

The test is not necessarily about copying to a different project. It is more a test about copying to a specific project.

danieljbruce committed:
Test for gax options in the test

Make sure that the gax options get passed down to the request layer.

danieljbruce committed:
TODO is done

The promise and callbacks follow the same pattern that is applied everywhere.

danieljbruce committed:
Change comment slightly

It is actually more of a check than an action that ensures something.

danieljbruce committed:
Eliminate TODO

The id is required. It can’t be generated automatically by the server. It needs to be provided.

danieljbruce committed:
Fix documentation for source function

Add description for the parameters and explain what the function does.

danieljbruce committed:
Move copy backup code to proper place

The tests were originally placed in a block at the very beginning so that they were easier to work through. This change moves them to the proper place.

danieljbruce committed:
Use cluster instead of parent

cluster should be used instead of parent to specify the destination cluster.

danieljbruce committed:
Change parent to cluster

cluster name should be used instead of parent to match the design doc.

danieljbruce committed:
A couple cleanup changes

Id is required so no question mark needed. Also inline the function that replaces the project name.

danieljbruce committed:
Use parent instead of cluster

This test was accidentally changed by a refactor and should not change further.

danieljbruce committed:
Eliminate unused time references

Variables were created so that time can be used in a block of tests. That block of tests were moved so these. variables are no longer needed.

danieljbruce committed:
This test should not have changed

Use parent instead of cluster because that is what it was before.

danieljbruce committed:
Eliminate unused import

CreateBackupConfig is not used anywhere.

danieljbruce committed:
Remove assert statement that is not needed

danieljbruce committed:
Move two lines of code to location used

General cleanup for readability

danieljbruce committed:
Improve readability

Reduce number of lines required in testWithExpiryTimes function. This makes the function easier to read.

danieljbruce committed:
Inline variables instead of using them explicitly

Don’t use the variables explicitly. Inline them instead.

danieljbruce committed:
Generate the backup id inline

An extra variable is not needed for this.

danieljbruce committed:
readability - indent code so it is separate

Also inline the id generation as it is not used elsewhere.

danieljbruce committed:
Indent block to separate check

Separate expiry time check from the rest of the code.

danieljbruce committed:
indent cluster creation, expiry time check

Indenting the code makes it easier to see variable relationships

danieljbruce committed:
Another test cleanup

Inline Bigtable instance creation options. Indent code for checking expiry time.

danieljbruce committed:
Add comment

Comment is for test that copies a table onto another project.

danieljbruce committed:
Indent code for table data insertion

Indenting this code shows it is not referenced later.

danieljbruce committed:
Rename variables to distinguish operations

Differentiate copy operation from create operation.

danieljbruce committed:
Eliminate unused imports

danieljbruce committed:
should be lower case

danieljbruce committed:
Change comment to respect alignment

danieljbruce committed:
Inline callback, config and cluster

Inline variables for better readability

danieljbruce committed:
Remove unused operation

danieljbruce committed:
Merge branch 'main' into copy-backup-2-relevant-changes-only

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gcf-owl-bot[bot] committed:
Merge branch 'copy-backup-2-relevant-changes-only' of https://github.com/danieljbruce/nodejs-bigtable into copy-backup-2-relevant-changes-only

danieljbruce committed:
Delete an instance

Delete the instance after it is used just like the other test.

danieljbruce committed:
Merge branch 'copy-backup-2-relevant-changes-only' of https://github.com/danieljbruce/nodejs-bigtable into copy-backup-2-relevant-changes-only

danieljbruce committed:
Inline generate id

A generateId function I need to inline.

kolea2 commented on src/utils/cluster.ts:
```diff
@@ -163,4 +165,10 @@ export class ClusterUtils {
     };
   }
+
+  static setExpiryTime(backup: ModifiableBackupFields) {
```
rename - `setBackupExpiryTime`

danieljbruce replied:
Done

danieljbruce committed:
Add second project to config

Import the second project to use in the integration tests

danieljbruce committed:
Rename the function to setBackupExpiryTime

Rename this function to be more specific.

danieljbruce committed:
Replace the project for backup path

Currently this test is too specific. It does not work if a specific project is specified in the client so we must only check the rest of the backup path.

danieljbruce commented:
https://github.com/googleapis/nodejs-bigtable/pull/1372 will address the failing test in the system tests.

danieljbruce committed:
Merge branch 'main' into copy-backup-2-relevant-changes-only

danieljbruce committed:
Run the linter

Linter removes unnecessary comma.

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

kolea2 commented on src/utils/cluster.ts:
```diff
@@ -163,4 +165,10 @@ export class ClusterUtils {
   }
+
+  static setBackupExpiryTime(backup: ModifiableBackupFields) {
+    if (backup.expireTime instanceof Date) {
```
should we throw if the condition is not met here?

danieljbruce replied:
No. This function is just pulling out the exact code that is being used in createBackup at https://github.com/googleapis/nodejs-bigtable/pull/1153/files#diff-9ce792749808a4cb9055a011d90ec2c35709e9deeee913947cb43f7ca394c911L317-L321 as this code is now used twice to parse a user provided date. backup.expireTime is of BackupTimestamp type and this means it can be a Date, PreciseDate or google.protobuf.ITimestamp as you can see by the definition in https://github.com/googleapis/nodejs-bigtable/blob/d30e5133a55e1b8f269b3aa6a69b49415361ba34/src/backup.ts#L44. Throwing an exception here would not let users provide a google.protobuf.ITimestamp value for their request.

kolea2 replied:
I think this method is misleading then. If I call `setBackupExpiryTime` and I pass in a non Date object (like an `ITimestamp` as you mention), there is no information to me that I am sending in something invalid, and nothing is getting set in that case. Can we rename the method to something more specific to the use case, or not extract it out?

danieljbruce replied:
I changed the name to formatBackupExpiryTime. We should still extract it out so that we don't repeat code. Feel free to let me know if you have a different preferred name in mind.

kolea2 replied:
LGTM

cindy-peng commented on system-test/bigtable.ts:
```diff
@@ -1414,6 +1421,268 @@ describe('Bigtable', () => {
+            expireTime: copyExpireTime,
+          },
+          INSTANCE
+        );
```
Delete the instance here after testing?

danieljbruce replied:
This instance is already deleted in https://github.com/googleapis/nodejs-bigtable/blob/d30e5133a55e1b8f269b3aa6a69b49415361ba34/system-test/bigtable.ts#L98-L115.

danieljbruce committed:
Merge branch 'main' into copy-backup-2-relevant-changes-only

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-bigtable into copy-backup-2-relevant-changes-only

danieljbruce committed:
Merge branch 'copy-backup-2-relevant-changes-only' of https://github.com/danieljbruce/nodejs-bigtable into copy-backup-2-relevant-changes-only

kolea2 commented on src/backup.ts:
```diff
@@ -37,6 +37,15 @@ import {
+export type CopyBackupCallback = GenericBackupCallback<Operation>;
+export interface CopyBackupConfig extends ModifiableBackupFields {
+  cluster: Cluster;
+  gaxOptions?: CallOptions;
```
rename to `callOptions`?

kolea2 commented on src/backup.ts:
```diff
@@ -243,6 +253,45 @@ Please use the format 'my-backup' or '${cluster.name}/backups/my-backup'.`);
   }
 
+  /**
+   * Copy this backup
```
"Copy this backup..." to where? please add to the documentation

danieljbruce replied:
I added a paragraph here. Some functions in the client library have a long explanation of what they do and some don't, but I extended this one so its purpose is more clear.

kolea2 commented on src/backup.ts:
```diff
@@ -243,6 +253,45 @@ Please use the format 'my-backup' or '${cluster.name}/backups/my-backup'.`);
+   * Copy this backup
+   *
+   * @param {CopyBackupConfig} [config] The config that specifies all of the
+   * information about the destination backup.
```
what is the "destination backup"? Is that the backup that will be copied, or the backup to be copied? Please add some more info here.

danieljbruce replied:
I added some more information here about this parameter.

kolea2 reviewed (commented):
added some initial comments, need to come back to review the tests

danieljbruce committed:
Update comments for the copy function

This function needs to be more clear and explain exactly what it does without ambiguity.

danieljbruce committed:
Rename function that sets correct value of expiry

expiryTime should be in the right format so rename function to more clearly indicate what it is doing.

danieljbruce committed:
Add an environment variable for second project

The variable for the second project is needed for the the integration tests to run properly for the copy backup test that copies to another project.

danieljbruce committed:
Add log for second project

See what kokoro prints out.

danieljbruce committed:
Remove the console log

This was just for testing. Remove this now.

danieljbruce committed:
Merge branch 'main' into copy-backup-2-relevant-changes-only

danieljbruce committed:
Merge branch 'main' into copy-backup-2-relevant-changes-only

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

kolea2 commented on system-test/bigtable.ts:
```diff
@@ -1414,6 +1421,268 @@ describe('Bigtable', () => {
+          assert.deepStrictEqual(backup.expireDate, sourceExpireTime);
+        }
+        // Create client, instance, cluster for second project
+        const bigtable = new Bigtable(
```
nit - I would rename this to something that signifies it's the second project. Maybe `bigtableSecondaryProject`?

kolea2 replied:
or `bigtableDestinationProject`?

danieljbruce replied:
Done

kolea2 commented on system-test/bigtable.ts:
```diff
@@ -1414,6 +1421,268 @@ describe('Bigtable', () => {
+            ? {projectId: process.env.GCLOUD_PROJECT2}
+            : {}
+        );
+        const instanceId = generateId('instance');
```
nit - I think we can inline this to the line below

danieljbruce replied:
Done

kolea2 commented on system-test/bigtable.ts:
```diff
@@ -1414,6 +1421,268 @@ describe('Bigtable', () => {
+            : {}
+        );
+        const instanceId = generateId('instance');
+        const instance = bigtable.instance(instanceId);
```
as above, I would rename this to show it's the second project instance

danieljbruce replied:
Done

danieljbruce committed:
Rename bigtable to bigtableSecondaryProject

danieljbruce committed:
Merge branch 'copy-backup-2-relevant-changes-only' of https://github.com/danieljbruce/nodejs-bigtable into copy-backup-2-relevant-changes-only

danieljbruce committed:
PR follow-ups - inline code and rename variables

danieljbruce committed:
Merge branch 'main' into copy-backup-2-relevant-changes-only

leahecole commented on system-test/bigtable.ts:
```diff
@@ -1414,6 +1421,269 @@ describe('Bigtable', () => {
+        assert.strictEqual(fetchedNewBackup.id, id);
+        assert.strictEqual(fetchedNewBackup.name, backupPath);
+        // Delete the copied backup
+        await config.cluster.backup(fetchedNewBackup.id).delete();
```
If this test fails, do you need to worry about backups being leftover? Or is the entire resource destroyed no matter what?

danieljbruce replied:
I added better cleanup for the backups both created from scratch and copied.

leahecole commented on test/backup.ts:
```diff
@@ -223,6 +225,68 @@ describe('Bigtable/Backup', () => {
 
+  describe('copy', () => {
+    beforeEach(() => {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
```
A few questions
1) What's the reasoning behind the eslint-disable-next-line plus the ts-ignore? In my mind, they kind of cancel each other out, but I'm sure that there's a deeper reason I just don't know here 🙂 
2) Is there a way we can not ignore the linter here?
3) if we're ignoring the linter, can we get away with only the eslint-disable-next-line and whatever the error is it's throwing? Or some kind of ts-expect-error? (I can't remember our exact settings for both eslint and the ts checker)

danieljbruce replied:
These comments may have been leftover from a copy/paste. They are not needed anymore.

danieljbruce committed:
Remove unused import

danieljbruce commented on src/cluster.ts:
```diff
@@ -314,11 +314,7 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
-        reqOpts.backup.expireTime
-      ).toStruct();
-    }
+    ClusterUtils.formatBackupExpiryTime(reqOpts.backup);
```
Yes!

danieljbruce committed:
Add comments for expire time

danieljbruce commented on system-test/bigtable.ts:
```diff
@@ -1414,6 +1421,269 @@ describe('Bigtable', () => {
     });
+    describe('copying backups', () => {
+      const sourceExpireTimeMilliseconds =
+        PreciseDate.now() + (8 + 300) * 60 * 60 * 1000;
```
My local system tests are hitting quotas so it's hard to recall exactly by testing it out. But if I recall correctly, the create backup expire time has to be sufficiently far ahead of the current time and the copied backup expire time has to be sufficiently ahead of the create backup expire time to avoid server errors specific to the copy backup feature. I added a comment for this.

leahecole replied:
Thanks for the comment! Makes sense as to why it needs to be a large number. My remaining confusion is why the arithmetic needs to be there rather than adding a fixed integer to PreciseDate.now()?

danieljbruce replied:
To your point, I think we can change this to `308 * 60 * 60 * 1000` and `608 * 60 * 60 * 1000`, but I want to keep the 60 * 60 * 1000 in there because it is beneficial to the reader to know that 308 is the number of hours (milliseconds/second * seconds/minute * minutes/hour).

leahecole replied:
ah, thank you! I agree with the 60 * 60 * 1000 making sense. I made a suggestion to modify the comment

danieljbruce replied:
Suggestions applied

danieljbruce committed:
Add better garbage cleanups of the backups

danieljbruce committed:
These comments are not needed anymore.

danieljbruce committed:
Remove only

leahecole commented on system-test/bigtable.ts:
```diff
@@ -1414,6 +1421,296 @@ describe('Bigtable', () => {
       Object.keys(policy).forEach(key => assert(key in updatedPolicy));
     });
+    describe('copying backups', () => {
+      // The server requires the source backup time to be sufficiently ahead
+      // of the current time to avoid an error.
+      // The server requires the copy backup time to be sufficiently ahead of
+      // the create backup time to avoid an error.
+      const sourceExpireTimeMilliseconds =
+        PreciseDate.now() + (8 + 300) * 60 * 60 * 1000;
```
```suggestion
      // The server requires the copy backup time to be sufficiently ahead of
      // the create backup time to avoid an error.
      // Set it to 308 hours ahead
      const sourceExpireTimeMilliseconds =
        PreciseDate.now() + (8 + 300) * 60 * 60 * 1000;
```

leahecole commented on system-test/bigtable.ts:
```diff
@@ -1414,6 +1421,296 @@ describe('Bigtable', () => {
+      const sourceExpireTimeMilliseconds =
+        PreciseDate.now() + (8 + 300) * 60 * 60 * 1000;
+      const sourceExpireTime = new PreciseDate(sourceExpireTimeMilliseconds);
+      const copyExpireTimeMilliseconds =
+        PreciseDate.now() + (8 + 600) * 60 * 60 * 1000;
```
```suggestion
      // 608 hours ahead of now, 300 hours ahead of sourceExpireTimeMilliseconds
      const copyExpireTimeMilliseconds =
        PreciseDate.now() + (8 + 600) * 60 * 60 * 1000;
```

leahecole reviewed (approved):
approving once suggestions with comments are added and with the assumption that the failing conformance tests are unrelated

danieljbruce committed:
Update system-test/bigtable.ts

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

danieljbruce committed:
Update system-test/bigtable.ts

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>
